lib/base/io-engine.hpp

```diff
@@ -174,40 +174,24 @@ class Timeout : public SharedObject
 	Timeout(boost::asio::io_context& io, Executor& executor, TimeoutFromNow timeoutFromNow, OnTimeout onTimeout)
 		: m_Timer(io)
 	{
-		Ptr keepAlive (this);
```

Comment:
While you're at it, why not adjust the parameters as well? For example, does the `onTimeout` handler have to be copied each time?

Reply:
Great idea! I also had it myself today, but I've already requested PR review.

Reply:
The parameters are still passed by copy! Why can't they be passed by reference?

Reply:
onTimeout is constructed in-place from a caller's lambda and then **moved** into ours. It can't get better. If I passed it by ref, I'd effectively copy it into the lambda and throw the original away.

Reply:
> onTimeout is constructed in-place from a caller's lambda

That doesn't mean that it won't get copied, but that's the least we can get out of it, I guess!
```cxx
#include <iostream>
#include <string>

struct test {
    int x_{0};
    std::string identity_;
    
    test(int x): x_(x), identity_("original") {}
    test(const test& other): x_(other.x_), identity_("Copy-"+std::to_string(other.x_)) {}

    void operator()() const {
        std::cout << "Identity: " << identity_ << " - X is : " << x_ << std::endl;
    }
};

template<class T>
void run(T f) {
    auto inner([f = std::move(f)] {
        f();
    });

    inner();
}

int main()
{
    run(test(1));

    return 0;
}

...
Identity: Copy-1 - X is : 1
```

Reply:
> That doesn't mean that it won't get copied,

Absolutely! Especially if you (literally) provide just a copy constructor instead of the _rule of five_, std::move() silently fails. But it's the best we can do _for the case_ T is move-aware.